engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/ThrowPanicNode.java

```diff
@@ -1,16 +1,81 @@
 package org.enso.interpreter.node.expression.builtin.error;
 
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.BranchProfile;
 import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.callable.atom.Atom;
 import org.enso.interpreter.runtime.error.PanicException;
 import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(
     type = "Panic",
     name = "throw",
     description = "Throws a new Panic with given payload.")
-public class ThrowPanicNode extends Node {
-  Stateful execute(Object _this, Object payload) {
+public abstract class ThrowPanicNode extends Node {
+  static ThrowPanicNode build() {
+    return ThrowPanicNodeGen.create();
+  }
+
+  abstract Stateful execute(Object _this, Object payload);
+
+  Context getContext() {
+    return Context.get(this);
+  }
+
+  @Specialization(guards = {"payload.getConstructor() == getContext().getBuiltins().caughtPanic()"})
+  Stateful doCaughtPanic(
+      Object _this,
+      Atom payload,
+      @CachedLibrary(limit = "5") InteropLibrary interopLibrary,
+      @Cached BranchProfile typeErrorProfile) {
+    // Note [Original Exception Type]
+    Object originalException = payload.getFields()[1];
+    if (interopLibrary.isException(originalException)) {
+      try {
+        throw interopLibrary.throwException(originalException);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(
+            "Impossible, `isException` returned true for `originalException`.");
```

Comment:
nit in wording
```suggestion
            "Unexpected true `isException` condition for `originalException`");
```
If something is impossible then there is no way it happens :) 
But I noticed this being used in other places so feel free to ignore.
Also notice that sometimes exceptions end with a dot and sometimes not. My preference is for the latter but as long as things are consistent, idc

Reply:
Hmmm, fair point. Well I guess the idea is that it is 'impossible under the assumption that the functions we use abide by their contracts'.

Maybe worth considering rephrasing but as you pointed out - it is used for many places, so I'd prefer to be consistent - we may change all of them in some separate PR though if we agree that different phrasing would be better. On the other hand, if the condition indeed is impossible it may not be worth caring about it too much :sweat_smile: